### DISCUSSION TITLE
Record name validation

### DISCUSSION THREAD

**Discussion by jean1** [auroraresearchlab/netbox-dns#267] [Ideas]
Currently, in netbox-dns, record name (and zone name) have no validation except on string length, whereas Netbox itself validates FQDN with a regexp. The lack of input validation is fragile if you want to use netbox-dns records to directly generate a zone file in Bind: the zone won't load. Record have to follow a certain number of rules. 

Each part of a name (the strings between the dots) is called a label. 
According to RFC 1035, 2.3.1., a label:
- doesn't start with a dash "-", 
- begin with a letter or a number,
- contains chars among: alphanumeric characters "a-zA-Z0-9", dash "-"
Also, a label can't exceed 63 characters (RFC 1035, 2.3.1. again: " There are also some restrictions on the length.  Labels must be *63* characters or less")

Other things to validate:
- name total length (concatenation of all labels and dots) < 255 chars

(the following things may be specific to bind zone file format)
- A dot "." is allowed in a name, "abc.def" in zone "example.com",
- You can have a relative ("abc") or an absolute name ending with a dot  ("abc.example.com."),
- A name has no 2 consecutive dots, ".." ,
- The special symbol "@" (alone) represents the zone name (the origin),
- Some record (type SRV, TXT but not A and AAAA)  may start with an underscore such as "_tcp".

My opinion is, if this is the central source of truth for your DNS records, allowing incorrectly formatted names and catching error later when generating zone file is undesirable.  

I agree though  that which rules to apply is not obvious (and blindly applying most of the rules would rule out IDN without an external conversion tool).

What would be an acceptable validation level ? Should this be opened as an issue ?

**Comment by peteeckel** [comment]
Hi @jean1, thanks for raising this!

I must admit I was a bit surprised at first, but after looking into the code I found that in fact we only validate the RDATA, not the name of the record, which is a stupid oversight.

As for the scope of validation, I'd use RFC 1035 and friends for that, keeping implementation specific stuff out as much as possible. There might be some exceptions (for example NetBox DNS uses @ for the zone name itself, which is somewhat historical and possibly going to go away at some point. Also, names with dots in them are fairly common, but I'm not sure whether they are specified in any RFC), but generally I would prefer to move along the lines of what the RFCs permit.

Since this lack of validation might cause malfunctions, please open an issue for it. I'll look into it as soon as possible.

**Reply by peteeckel** [reply]
@jean1: I did some RFC research on this, and there are some quite surprising facts ... bottom line: It's not that easy :-)

[RFC1035, Section 2.3.1](https://datatracker.ietf.org/doc/html/rfc1035#section-2.3.1) on names:
```
The following syntax will result in fewer problems with many
applications that use domain names (e.g., mail, TELNET).

<domain> ::= <subdomain> | " "
<subdomain> ::= <label> | <subdomain> "." <label>
<label> ::= <letter> [ [ <ldh-str> ] <let-dig> ]
<ldh-str> ::= <let-dig-hyp> | <let-dig-hyp> <ldh-str>
<let-dig-hyp> ::= <let-dig> | "-"
<let-dig> ::= <letter> | <digit>
<letter> ::= any one of the 52 alphabetic characters A through Z in
upper case and a through z in lower case
<digit> ::= any one of the ten digits 0 through 9

Note that while upper and lower case letters are allowed in domain
names, no significance is attached to the case.  That is, two names with
the same spelling but different case are to be treated as if identical.

The labels must follow the rules for ARPANET host names.  They must
start with a letter, end with a letter or digit, and have as interior
characters only letters, digits, and hyphen.  There are also some
restrictions on the length.  Labels must be 63 characters or less.
```
This would in fact imply, for instance, that a label **must** start with a letter, then letters, digits and hyphens are allowed, and it must end with a letter or digit. So far, so good, but look at the `in-addr.arpa` and `ip6.arpa` zones and you'll find that this can't be correct, at least not anymore.

A bit later in [RFC 2181, Section 11](https://www.rfc-editor.org/rfc/rfc2181.html#section-11) things look completely different:
```
   The DNS itself places only one restriction on the particular labels
   that can be used to identify resource records.  That one restriction
   relates to the length of the label and the full name.  The length of
   any one label is limited to between 1 and 63 octets.  A full domain
   name is limited to 255 octets (including the separators).  The zero
   length full name is defined as representing the root of the DNS tree,
   and is typically written and displayed as ".".  Those restrictions
   aside, any binary string whatever can be used as the label of any
   resource record.  Similarly, any binary string can serve as the value
   of any record that includes a domain name as some or all of its value
   (SOA, NS, MX, PTR, CNAME, and any others that may be added).
```
Whoops. That's something entirely different.

However, this raises another interesting issue: The RFCs refer to the length of labels and domain names in terms of octets, and to the wire format and not to the textual representation in zone files. Many characters that are by the definition in RFC 2181 legal need to be escaped in zone files, which makes their textual representation longer than they actually are on the wire. So before calculating the length, names need to be converted to wire format to determine their actual lengths. 

The good news is that probably we don't need to take care of that at all (no regular expressions etc.) and have `dnspython` do all the work for us.

```
>>> import dns
>>> name='thislabelnameisfartoolongtobepermittedbyrfc1035andnowidonthaveanyideawhattowritesoijustaddsomenonsense.example.org'
>>> dnsname=dns.name.from_text(name)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.10/site-packages/dns/name.py", line 959, in from_text
    return Name(labels)
  File "/usr/local/lib/python3.10/site-packages/dns/_immutable_ctx.py", line 41, in nf
    f(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/dns/_immutable_ctx.py", line 64, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/dns/name.py", line 327, in __init__
    _validate_labels(self.labels)
  File "/usr/local/lib/python3.10/site-packages/dns/name.py", line 286, in _validate_labels
    raise LabelTooLong
dns.name.LabelTooLong: A DNS label is > 63 octets long.
```

So, in a first approach, I suggest I'll just run record and zone names through the `dnspython` validation and that's it - trusting that `dnspython` is maintained well and implements the (few) restrictions the RFCs impose.
